Toplean_Alexandru/week-02/Demo/Program.cs

```diff
@@ -0,0 +1,125 @@
+            var serviceProvider = serviceCollection.BuildServiceProvider();
+
+            // from restaurantResult in RestaurantDomain.CreateRestaurant("McDonalds") //Exception -> RestaurantErrorCode.RestaurantExists
+            // from restaurantResult in RestaurantDomain.CreateRestaurant("McDonalds►") //Exception -> RestaurantErrorCode.IllegalCharacters
+            // from restaurantResult in RestaurantDomain.CreateRestaurant("") //Exception -> RestaurantErrorCode.EmptyField
+            var expr =
+                from restaurantResult in RestaurantDomain.CreateRestaurant("mcdonalds")
+                let restaurant = (restaurantResult as RestaurantCreated)?.Restaurant
+                select restaurantResult;
+
+            var interpreter = new LiveInterpreterAsync(serviceProvider);
+            var result = await interpreter.Interpret(expr, Unit.Default);
+
+            // dynamic keyword needed to get the result type, error otherwise
+            dynamic resultType = result.GetType();
+
+            // If the restaurant was successfully created
+            if (resultType.Name == "RestaurantCreated")
+            {
+                // Restaurant was successfully created
+                var resultClass = result as RestaurantCreated;
+            } // End RestaurantCreated If
+            else
+            {
+                // Treating all the possible errors
+                var resultClass = result as RestaurantNotCreated;
+                Console.Write("Error! Restaurant was not created because the ");
+                switch (resultClass.Reason)
+                {
+                    case RestaurantErrorCode.NameTooLong:
+                        {
+                            Console.WriteLine($"restaurant name was too long. Please use a name shorter than 256 characters! Error code {resultClass.Reason}");
+                            break;
+                        }
+                    case RestaurantErrorCode.IllegalCharacters:
+                        {
+                            Console.WriteLine($"restaurant name had illegal characters. Error code {resultClass.Reason}");
+                            break;
+                        }
+                    case RestaurantErrorCode.RestaurantExists:
+                        {
+                            Console.WriteLine($"restaurant already exists in the system. Error code {resultClass.Reason}");
+                            break;
+                        }
+                    case RestaurantErrorCode.EmptyField:
+                        {
+                            Console.WriteLine($"restaurant name was empty. Error code {resultClass.Reason}");
+                            break;
+                        }
+                }
+            }
+
+            var restaurantCreated = result as RestaurantCreated; // Assuming we have the createdRestaurant or using an existing restaurant
+
+            // Trying to create a menu
```

Comment:
I think that this is a very good start to understand the pattern matching properties of that result. What happened here is that you actually tried to avoid the fact that we are hiding all the useful data behind that abstraction (ICreateRestaurantResult). 

Would you please try to refactor this code using the Match extension methods? You already have them defined and invoked in the following line:

```csharp
  var finalResult = result.Match<bool>(OnRestaurantCreated, OnRestaurantNotCreated);
```
...you just have to fill in the implementations for OnRestaurantCreated and OnRestaurantNotCreated.